### PR TITLE
Standardize NaNs in the interpreter, when there is nondeterminism

### PR DESCRIPTION
**Pull request by kripken**
Specifically, pick a positive sign when both are valid outputs. This is enough
to fix a testcase the fuzzer found:

https://github.com/WebAssembly/binaryen/pull/3289#issuecomment-718784793

See the comment in the source for full details.

I *think* this is right, but this stuff is tricky. @sunfishcode @binji maybe you can
take a look?

### CODE REVIEW

File: src/wasm/literal.cpp [hunk 1]
```diff
@@ -837,16 +837,56 @@ Literal Literal::demote() const {
...
+// 0/0, then we should not do any standardization).
+template<typename T> static Literal standardizeNaN(T left, T right, T result) {
+  if (!std::isnan(left) && !std::isnan(right)) {
+    // No NaN propagation, leave the result.
+    return Literal(result);
+  }
+  // In the presence of nondeterminism, the sign is arbitrary. Be positive
+  // consistently.
+  return Literal(std::copysign(result, T(1)));
+}
```
Comment (MaxGraey) [inline src/wasm/literal.cpp:872]: what about canonicalize this to arithmetic if `left` or `right` non-canonical? Something like this:
 ```cpp
template<typename T> static Literal standardizeNaN(Literal left, Literal right, T result) {
  if (left.isNaN()) {
    if (Literal::NaNPayload(left.getFloat())) { // non-canonical NaN
      if (std::signbit(left.getFloat())) left = left.neg(); // remove sign bit
      return Literal(Literal::setQuietNaN(left));// return arithmetical NaN
    }
    return left; // left is canonical NaN
  }
  if (right.isNaN()) {
    if (Literal::NaNPayload(right.getFloat())) { // non-canonical NaN
      if (std::signbit(right.getFloat())) right = right.neg(); // remove sign bit
      return Literal(Literal::setQuietNaN(right));// return arithmetical NaN
    }
    return right; // right is canonical NaN
  }
  return Literal(result);
}
```

Comment (kripken) [reply src/wasm/literal.cpp:872]: @MaxGraey Perhaps. For now I hope this PR is enough. I'd rather not add something until it's clear we need it. But we may need it later.

**Comment by kripken** [comment]
@tlively are the NaN propagation rules different for SIMD? This fails a spec test,
```wat
(assert_return (invoke "f32x4.add" (v128.const f32x4 nan -nan      infinity 42)
                                   (v128.const f32x4  42  infinity infinity  1))
                                   (v128.const f32x4 nan -nan      infinity 43))
```
That test seems to assume  that `-nan + infinity = -nan`, that is, that the sign of the nan is unchanged.

**Comment by tlively** [comment]
No, the NaN propagation rules are the same for SIMD. The SIMD "spec tests" here in Binaryen are actually just ad hoc tests I've written for the interpreter, so they're by no means authoritative when it comes to floating point details.

**Comment by kripken** [comment]
I see, thanks @tlively . So sounds like we just need to update them, I did that now.

File: src/wasm/literal.cpp (hunk 1 above)
Comment (kripken) [reply src/wasm/literal.cpp]: It turns out we do need more, as the tests seem to hit nondeterminism between my machine and CI :wink: 

To solve that, just always return a simple canonical NaN that is positive, which I believe is always valid.

**Comment by binji** [comment]
It seems like there is a simpler solution here, but perhaps I'm misunderstanding what this code is trying to accomplish. The way I think of it is this: if a NaN is produced, it is almost always non-deterministic. The only instructions where that is not true (AFAIK) are `fNN.abs`, `fNN.copysign`, `fNN.neg`, `fNN.const`, `fNN.reinterpret_iNN`, and `fNN.load`. So you can canonicalize a NaN after any other float-producing operation.

**Comment by kripken** [comment]
@binji You're right, I should have realized that initially. Ok, things are hopefully simple and correct now...

**Comment by kripken** [comment]
Ok, this should be ready for review now - passes tests and does what @binji said.

File: src/wasm/literal.cpp [hunk 2]
```diff
@@ -837,16 +837,61 @@ Literal Literal::demote() const {
...
+// something. In other words, if the valid output of an expression is a set of
+// valid NaNs, and after optimization the output is still that same set, then
+// the optimization is valid. And if the interpreter picks the same NaN in both
+// cases from that identical set then nothing looks wrong to the fuzzer.
+//
+// Note that this function must receive both inputs to the operation, so that it
+// can tell if a NaN existed beforehand (if a NaN was generated just now, as in
+// 0/0, then we should not do any standardization).
+template<typename T> static Literal standardizeNaN(T left, T right, T result) {
+  if (!std::isnan(left) && !std::isnan(right)) {
```
Comment (binji) [inline src/wasm/literal.cpp:865]: Is this necessary? It seems like you can check whether the just the `result` is a NaN value. You mention this above: "if a NaN was generated just now, as in 0/0, then we should not do any standardization", but I don't see why that would be true. It's still possible for that operation to produce a non-deterministic NaN.

Comment (kripken) [reply src/wasm/literal.cpp]: Thanks, I see you're right, I missed this in the spec: `including the case that there are no NaN inputs`. Updated.

File: test/passes/optimize-instructions_fuzz-exec.txt [hunk 3]
```diff
@@ -0,0 +1,255 @@
+[fuzz-exec] calling test32
+[LoggingExternalInterface logging nan:0x400000]
+[LoggingExternalInterface logging nan:0x400000]
+[LoggingExternalInterface logging nan:0x400000]
+[LoggingExternalInterface logging nan:0x400000]
+[LoggingExternalInterface logging nan:0x7fff82]
+[LoggingExternalInterface logging -nan:0x7fff82]
+[LoggingExternalInterface logging -nan:0x7fff82]
```
Comment (tlively) [inline test/passes/optimize-instructions_fuzz-exec.txt:8]: It looks like there are some non-standard NaNs still floating around. Would it be worth standardizing these cases, too?

File: src/wasm/literal.cpp [hunk 4]
```diff
@@ -989,7 +1023,7 @@ Literal Literal::div(const Literal& other) const {
         case FP_ZERO:
           switch (std::fpclassify(lhs)) {
             case FP_NAN:
-              return Literal(setQuietNaN(lhs));
+              return standardizeNaN(lhs / rhs);
             case FP_ZERO:
               return Literal(
                 std::copysign(std::numeric_limits<float>::quiet_NaN(), sign));
```
Comment (tlively) [inline src/wasm/literal.cpp:1029]: Should this NaN be standardized as well?

File: test/passes/optimize-instructions_fuzz-exec.txt (hunk 3 above)
Comment (kripken) [reply test/passes/optimize-instructions_fuzz-exec.txt]: Fixed, thanks, those were min and max which were missed. There remains a non-standard NaN which is for copysign which is expected, as it does not propagate NaNs with the normal rules.

File: src/wasm/literal.cpp (hunk 4 above)
Comment (kripken) [reply src/wasm/literal.cpp]: Done.

**Review by tlively — APPROVED** [review]

**Pull request merged**